Studio: keymap subsystem (#2268)
zmkfirmware/zmk: docs/docs/development/studio-setup.md

| Height     | int (>0) | Key(cap) height                      | [centi-](https://en.wikipedia.org/wiki/Centi-)"keyunit" |
| X          | uint     | Key X position (top-left point)      | [centi-](https://en.wikipedia.org/wiki/Centi-)"keyunit" |
| Y          | uint     | Key Y position (top-left point)      | [centi-](https://en.wikipedia.org/wiki/Centi-)"keyunit" |
| Rotation   | int      | Key rotation (positive => clockwise) | [centi-](https://en.wikipedia.org/wiki/Centi-)degree    |
| Rotation X | int      | Rotation origin X position           | [centi-](https://en.wikipedia.org/wiki/Centi-)"keyunit" |
| Rotation Y | int      | Rotation origin Y position           | [centi-](https://en.wikipedia.org/wiki/Centi-)"keyunit" |

:::note
You can specify negative values in devicetree using parentheses around it, e.g. `(-3000)` for a 30 degree counterclockwise rotation.
:::

## Header Include

To pull in the necessary definition for creating physical layouts, a new include should be added to the top of the devicetree file:

```
#include <physical_layouts.dtsi>
```

## Example

Here is an example physical layout for a 2x2 macropad:

> ```

Reviewer: ````suggestion
```dts
````

However, keys between layouts might not be in exactly the same positions, in which case a position map can be used. The position map includes a sequence for every relevant layout, and the corresponding entries in `positions` property will be used to determine the mapping between layouts. By default, the physical attribute matching behavior will be used as a fallback for positions not specified in the map, but the `complete` property can be added to the map to specify that no matching fallback should occur.

## Examples

### Basic Map

For example, the following position map correctly maps the 5-column and 6-column Corne keymap layouts.

```dts
    foostan_corne_position_map {
        compatible = "zmk,physical-layout-position-map";

        complete;

        twelve {
            physical-layout = <&foostan_corne_6col_layout>;
            positions
                = < 1  2  3  4  5  6  7  8  9 10>
                , <13 14 15 16 17 18 19 20 21 22>
                , <25 26 27 28 29 30 31 32 33 34>

>                 , <      36 37 38 39 40 41      >;

Reviewer: I think this is missing the first and last columns by accident; shouldn't it be the same as the value in the "full preserving map" example?

---
title: Studio Setup
---

:::warning[Alpha Feature]

> ZMK Studio support is in alpha at this point. Although our best efforts are being made to keep compatibility during active development, but is not guaranteed.

Reviewer: ```suggestion
ZMK Studio support is in alpha. Although best efforts are being made, keeping compatibility during active development is not guaranteed.
```

:::

This guide will walk you through enabling ZMK Studio support for a keyboard.

The main additional pieces needed for Studio support involve additional metadata needed in order
to properly to display the physical layouts available for the particular keyboard.

# Physical Layout Positions

> Physical layouts are described as part of the [New Shield Guide](./new-shield.mdx#physical-layouts), but additional possible property is not described there. The `keys` property is used to describe the physical attributes of each key position present in that layout, and are listed in the same order as keymap bindings, matrix transforms, etc. The properties available are:

Reviewer: ```suggestion
Physical layouts are described as part of the [new shield guide](./new-shield.mdx#physical-layouts) with the exception of the `keys` property that is required for Studio support. This is used to describe the physical attributes of each key position present in that layout and its items are listed in the same order as keymap bindings, matrix transforms, etc. The properties available are:
```

- Width
- Height
- X
- Y
- Rotation
- Rotation X
- Rotation Y

All X/Y/Width/Height values are in "key unit" units, with a 100th of a key unit resolution. Rotations are in degrees, with a 100th of a degree resolution. Due to the nature of devicetree, all values are stored in the devicetree without decimals as hundredths of a value, e.g. one key unit is represented as `100`, twenty degrees is represented as `2000`, etc.

> Here is an example physical layout for a 2x2 macropad:

Reviewer: I think we are missing a paragraph on rotation here, something like:

```suggestion
If the Rotation value is non-zero, the key will be rotated by that angle around the coordinates specified by Rotation X/Y values, after it is translated to X/Y coordinates. Rotation is clockwise for positive values and counter clockwise otherwise.

:::note
You can specify negative values in devicetree using parentheses around it, e.g. `(-3000)` for a 30 degree counter-clockwise rotation.
:::

Here is an example physical layout for a 2x2 macropad:
```

I also think the negative value pointer would be useful here.

```
    macropad_physical_layout: macropad_physical_layout {
        compatible = "zmk,physical-layout";
        display-name = "Macro Pad";

        keys  //                     w   h    x    y     rot    rx    ry
            = <&key_physical_attrs 100 100    0    0       0     0     0>
            , <&key_physical_attrs 100 100  100    0       0     0     0>
            , <&key_physical_attrs 100 100    0  100       0     0     0>
            , <&key_physical_attrs 100 100  100  100       0     0     0>
            ;
    };
```

# Position Map

> When switching between layouts with ZMK Studio, the keymap of the previously selected layout is used to populate the keymap in the new layout. To determine which keymap entry maps to which entry in the new layout, keys between the two layouts are matched that share the exact same physical attributes. However, in some cases, keys between layouts are not in exactly the same position.

Reviewer: ```suggestion
When switching between layouts with ZMK Studio, the keymap of the previously selected layout is used to populate the keymap in the new layout. To determine which keymap entry maps to which entry in the new layout, keys between the two layouts that share the exact same physical attributes are matched.
```

> For those cases, a position map is used. The position map includes a sequence for every relevant layout, and the corresponding entries in the `positions` will be used to determine the mapping.

Reviewer: ```suggestion
However, keys between layouts might not be in exactly the same positions, in which case a position map can be used. The position map includes a sequence for every relevant layout, and the corresponding entries in `positions` property will be used to determine the mapping between layouts.
```

For example, the following position map correctly maps the 5-column and 6-column Corne keymap layouts.

> ```
>     foostan_corne_position_map {
>         compatible = "zmk,physical-layout-position-map";
> 
>         complete;
> 
>         twelve {
>             physical-layout = <&foostan_corne_6col_layout>;
>             positions
>                 = <    1  2  3  4  5  6  7  8  9 10   >
>                 , <   13 14 15 16 17 18 19 20 21 22   >
>                 , <   25 26 27 28 29 30 31 32 33 34   >
>                 , <         36 37 38 39 40 41               >;
>         };
> 
>         ten {
>             physical-layout = <&foostan_corne_5col_layout>;
>             positions
>                 = < 0  1  2  3  4  5  6  7  8  9>
>                 , <10 11 12 13 14 15 16 17 18 19>
>                 , <20 21 22 23 24 25 26 27 28 29>
>                 ,       <30 31 32 33 34 35>;
>         };
>     };
> ```

Reviewer: ````suggestion
```dts
    foostan_corne_position_map {
        compatible = "zmk,physical-layout-position-map";

        complete;

        twelve {
            physical-layout = <&foostan_corne_6col_layout>;
            positions
                = < 1  2  3  4  5  6  7  8  9 10>
                , <13 14 15 16 17 18 19 20 21 22>
                , <25 26 27 28 29 30 31 32 33 34>
                , <      36 37 38 39 40 41      >;
        };

        ten {
            physical-layout = <&foostan_corne_5col_layout>;
            positions
                = < 0  1  2  3  4  5  6  7  8  9>
                , <10 11 12 13 14 15 16 17 18 19>
                , <20 21 22 23 24 25 26 27 28 29>
                , <      30 31 32 33 34 35      >;
        };
    };
```
````

> The first entries in the two mappings have values `1` and `0` respectively, which means that the zero-th entry in the 5-column will map to the first entry in the 6-column layout, the second entries show that the second 6-column keymap entry corresponds to the first 5-column entry, etc.

Reviewer: I think this is missing documenting the `complete;` property.

Author: Added documentation above the example.

:::

This guide will walk you through enabling ZMK Studio support for a keyboard.

The main additional pieces needed for Studio support involve additional metadata needed in order
to properly to display the physical layouts available for the particular keyboard.

# Physical Layout Positions

Physical layouts are described as part of the [new shield guide](./new-shield.mdx#physical-layouts) with the exception of the `keys` property that is required for Studio support. This is used to describe the physical attributes of each key position present in that layout and its items are listed in the same order as keymap bindings, matrix transforms, etc. The properties available are:

- Width
- Height
- X
- Y
- Rotation
- Rotation X
- Rotation Y

All X/Y/Width/Height values are in "key unit" units, with a 100th of a key unit resolution. Rotations are in degrees, with a 100th of a degree resolution. Due to the nature of devicetree, all values are stored in the devicetree without decimals as hundredths of a value, e.g. one key unit is represented as `100`, twenty degrees is represented as `2000`, etc.

Width/Height values should be positive and X/Y should be nonnegative.

> If the Rotation value is non-zero, the key will be rotated by that angle around the coordinates specified by Rotation X/Y values, after it is translated to X/Y coordinates. Rotation is clockwise for positive values and counterclockwise otherwise.

Reviewer 2: I think this would read nicer as a table. Something like Property - Description - Unit - Resolution?

> - Width
> - Height
> - X
> - Y
> - Rotation
> - Rotation X
> - Rotation Y
> 
> All X/Y/Width/Height values are in "key unit" units, with a 100th of a key unit resolution. Rotations are in degrees, with a 100th of a degree resolution. Due to the nature of devicetree, all values are stored in the devicetree without decimals as hundredths of a value, e.g. one key unit is represented as `100`, twenty degrees is represented as `2000`, etc.
> 
> Width/Height values should be positive and X/Y should be nonnegative.
> If the Rotation value is non-zero, the key will be rotated by that angle around the coordinates specified by Rotation X/Y values, after it is translated to X/Y coordinates. Rotation is clockwise for positive values and counterclockwise otherwise.

Reviewer 2: ```suggestion
| Property | Type | Description                                             | Unit |
| ---------| ---- | ------------------------------------------ | ------- |
| Width    | int (>0) | Key(cap) width                                    | [centi-](https://en.wikipedia.org/wiki/Centi-)"keyunit"|
| Width    | int (>0) | Key(cap) height                                 | [centi-](https://en.wikipedia.org/wiki/Centi-)"keyunit"|
| X           | uint         | Key X position (center point)            | [centi-](https://en.wikipedia.org/wiki/Centi-)"keyunit"|
| Y           | uint         | Key Y position (center point)            | [centi-](https://en.wikipedia.org/wiki/Centi-)"keyunit"|
| Rotation | int         | Key rotation (clockwise = positive)    | [centi-](https://en.wikipedia.org/wiki/Centi-)degree |
| Rotation X | int         | Rotation origin X position relative to key X position | [centi-](https://en.wikipedia.org/wiki/Centi-)"keyunit"|
| Rotation Y | int         | Rotation origin Y position relative to key Y position | [centi-](https://en.wikipedia.org/wiki/Centi-)"keyunit"|

```
Wrote out a (poorly formatted) example of what I'm thinking.

Author: Updated, but with some tweaks for accuracy.

Reviewer 2: The updated version reads to me as if the rotation origin positions are no longer relative to the key. "Rotation origin relative X position"? If they are no longer relative then the 0,0 case needs to be stated explicitly.

Author: They are not relative to the X/Y position, but absolute. I can include the term "absolute", if you think that's descriptive enough.

:::note
You can specify negative values in devicetree using parentheses around it, e.g. `(-3000)` for a 30 degree counterclockwise rotation.
:::

Here is an example physical layout for a 2x2 macropad:

```
    macropad_physical_layout: macropad_physical_layout {
        compatible = "zmk,physical-layout";
        display-name = "Macro Pad";

        keys  //                     w   h    x    y     rot    rx    ry

>             = <&key_physical_attrs 100 100    0    0       0     0     0>

Reviewer 3: `&key_physical_attrs` is from `physical_layouts.dtsi` and needs to be `# include`d. This is currently not documented.

Author: Added.

            , <&key_physical_attrs 100 100  100    0       0     0     0>
            , <&key_physical_attrs 100 100    0  100       0     0     0>
            , <&key_physical_attrs 100 100  100  100       0     0     0>
            ;
    };
```

> 

Reviewer 2: Might be nice to include an image render of the above here.

# Physical Layout Positions

Physical layouts are described as part of the [new shield guide](./new-shield.mdx#physical-layouts) with the exception of the `keys` property that is required for Studio support. This is used to describe the physical attributes of each key position present in that layout and its items are listed in the same order as keymap bindings, matrix transforms, etc. The properties available are:

| Property   | Type     | Description                          | Unit                                                    |
| ---------- | -------- | ------------------------------------ | ------------------------------------------------------- |
| Width      | int (>0) | Key(cap) width                       | [centi-](https://en.wikipedia.org/wiki/Centi-)"keyunit" |
| Height     | int (>0) | Key(cap) height                      | [centi-](https://en.wikipedia.org/wiki/Centi-)"keyunit" |
| X          | uint     | Key X position (top-left point)      | [centi-](https://en.wikipedia.org/wiki/Centi-)"keyunit" |
| Y          | uint     | Key Y position (top-left point)      | [centi-](https://en.wikipedia.org/wiki/Centi-)"keyunit" |

> | Rotation   | int      | Key rotation (positive => clockwise) | [centi-](https://en.wikipedia.org/wiki/Centi-)degree    |
> | Rotation X | int      | Rotation origin X position           | [centi-](https://en.wikipedia.org/wiki/Centi-)"keyunit" |
> | Rotation Y | int      | Rotation origin Y position           | [centi-](https://en.wikipedia.org/wiki/Centi-)"keyunit" |

Reviewer 2: ```suggestion
| Rotation   | int      | Key rotation (positive &rArr; clockwise) | [centi-](https://en.wikipedia.org/wiki/Centi-)degree    |
| Rotation X | int      | Rotation origin relative X position           | [centi-](https://en.wikipedia.org/wiki/Centi-)"keyunit" |
| Rotation Y | int      | Rotation origin relative Y position           | [centi-](https://en.wikipedia.org/wiki/Centi-)"keyunit" |
```

Reviewer 2: Really splitting up the conversation here, but I wanted to grab the rArr for you :sweat_smile: What happens when you rotate around 0,0 - I thought it would rotate around the key center, but if it rotates around the 0,0 of the keyboard then that's fine as-is

Author: Yeah, it's an absolute rotation point, which is useful for a few scenarios, e.g. rotating a whole half of a layout by assigning the same rotation, and rotation X/Y to all of the relevant key positions..
